### PR TITLE
[POC] ANDROID-10106 mistica icons generated using script

### PR DESCRIPTION
**Pull request by manolovn**
[Open Discussion]
I want to discuss in this PR your thoughts about this proposal. There are some problems with and I want to talk about the viability of the idea before a final implementation

### Topics to discuss about
- How to release `mistica-icons`
  - Along with mistica library
  - Dedicated github action as a independent artifact
- When to generate the icon set
  - On demand manually when needed
  - On every release
- Current implementation
  - Keep generating `ImageVector`
  - Generate `Composable`s to be used as components
- Impact in app size
- How to deal with the `svg-to-compose` dependency
  - Upload to our nexus repository
  - Include the jar directly in the module
  - Automatize in the script the download and installation of the artifact

### :goal_net: What's the goal?
Icons genetaed as VectorImage using an script based on svg-to-compose repository

### ☑️ Checks
- [ ] I updated the documentation (readmes, wikis, etc). If no need to update documentation explain why.
- [ ] Tested with dark mode.
- [ ] Tested with API 21.

### :test_tube: How can I test this?
- [ ] 🖼️ Screenshots/Videos
- [ ] Reviewed by Mistica design team

### CODE REVIEW

File: mistica-icons/generate-icons.main.kts [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env kotlin
+
+@file:Repository("https://jitpack.io")
+@file:Repository("https://maven.google.com")
+@file:Repository("https://jetbrains.bintray.com/trove4j")
+@file:Repository("file:///home/mvera/.m2/repository")
```
Comment (manolovn) [inline mistica-icons/generate-icons.main.kts:6]: the library used, svg-to-compose doen't appear in any public repository so for this poc I've installed it locally

Comment (gmerinojimenez) [reply mistica-icons/generate-icons.main.kts:6]: image

Maybe we can add it to libs folder?

File: mistica-icons/generate-icons.main.kts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+val git: Git = Git.cloneRepository()
+    .setURI(repository)
+    .setDirectory(repoCloneDir)
+    .call()
+git.checkout().setName("production").call()
+
+val iconsDir = File(repoCloneDir, "icons")
+
+// renaming to match to svg-to-compose
+iconsDir.walkTopDown().filter { it.extension == "svg" }
```
Comment (gmerinojimenez) [inline mistica-icons/generate-icons.main.kts:45]: just curious, why `walkTopDown()`?

File: mistica-icons/src/main/kotlin/com/telefonica/mistica/compose/__MisticaIcons.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.telefonica.mistica.compose
+
+import androidx.compose.ui.graphics.vector.ImageVector
+import com.telefonica.mistica.compose.misticaicons.AllIcons
+import com.telefonica.mistica.compose.misticaicons.Telefonica
+import kotlin.collections.List as ____KtList
+
+public object MisticaIcons
+
+private var __AllIcons: ____KtList<ImageVector>? = null
```
Comment (gmerinojimenez) [inline mistica-icons/src/main/kotlin/com/telefonica/mistica/compose/__MisticaIcons.kt:10]: autogenerated?

**Review by gmerinojimenez — APPROVED** [review]
Very nice!

File: mistica-icons/generate-icons.main.kts (hunk 1 above)
Comment (dpastor) [reply mistica-icons/generate-icons.main.kts:6]: Yes, we should if possible, in order to require the minimum setup.

Check this approach used in android-messenger -> https://github.com/Telefonica/android-messenger/tree/integration/mavenLibs

We can store maven artifacts (with associated info files) locally in the repo and use them directly.

File: mistica-icons/build.gradle [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+android {
+    compileSdk 31
+
+    defaultConfig {
+        minSdk 21
+        targetSdk 31
+
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
```
Comment (dpastor) [inline mistica-icons/build.gradle:13]: Not needed, no?

File: mistica-icons/build.gradle [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        consumerProguardFiles "consumer-rules.pro"
+    }
+
+    buildTypes {
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+        }
+    }
```
Comment (dpastor) [inline mistica-icons/build.gradle:22]: needed?

File: mistica-icons/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest package="com.telefonica.mistica.icons">
+
+</manifest>
```
Comment (dpastor) [inline mistica-icons/src/main/AndroidManifest.xml:4]:
```suggestion
<manifest package="com.telefonica.mistica.icons" />
```

File: mistica-icons/generate-icons.main.kts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+@file:DependsOn("org.eclipse.jgit:org.eclipse.jgit:3.5.0.201409260305-r")
+
+import br.com.devsrsouza.svg2compose.Svg2Compose
+import br.com.devsrsouza.svg2compose.VectorType
+import org.eclipse.jgit.api.Git
+import java.io.File
+
+fun File.makeDirs() = apply { mkdirs() }
+
+val buildDir = File("build/").makeDirs()
```
Comment (dpastor) [inline mistica-icons/generate-icons.main.kts:28]: used?

File: mistica-icons/src/main/kotlin/com/telefonica/mistica/compose/__MisticaIcons.kt (hunk 3 above)
Comment (manolovn) [reply mistica-icons/src/main/kotlin/com/telefonica/mistica/compose/__MisticaIcons.kt:10]: yup, the total number of icons generated is 1100+; for this PoC discussion I've just added one icon.
The total generation includes o2, telefonica and blau icons with the different groups: filled, outline and regular.
There is an important blocker point, after the generation the library compilation breaks with OOM error :(

File: mistica-icons/build.gradle (hunk 4 above)
Comment (manolovn) [reply mistica-icons/build.gradle:13]: nope, removed

File: mistica-icons/build.gradle (hunk 5 above)
Comment (manolovn) [reply mistica-icons/build.gradle:22]: nope, removed

File: mistica-icons/generate-icons.main.kts (hunk 7 above)
Comment (manolovn) [reply mistica-icons/generate-icons.main.kts:28]: nope, removed

File: mistica-icons/generate-icons.main.kts (hunk 2 above)
Comment (manolovn) [reply mistica-icons/generate-icons.main.kts:45]: not strong preference, just playing with the different methods; with walk or walkBottomUp the script works

**Pull request closed without merging**